internal/features/stacks/jobs/builtin_references.go

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package jobs
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl-lang/lang"
+	"github.com/hashicorp/hcl-lang/reference"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/terraform-ls/internal/features/stacks/state"
+	"github.com/zclconf/go-cty/cty"
+)
+
+var orchestrateContextScopeId = lang.ScopeId("orchestrate_context")
+
+// used in various places
+var changesAttributes = map[string]cty.Type{
+	"total":  cty.Number,
+	"add":    cty.Number,
+	"change": cty.Number,
+	"import": cty.Number,
+	"remove": cty.Number,
+	"move":   cty.Number,
+	"forget": cty.Number,
+	"defer":  cty.Number,
+}
+var changesType = cty.Object(changesAttributes)
+
+func builtinReferences(record *state.StackRecord) reference.Targets {
+	targets := make(reference.Targets, 0)
+
+	if record == nil {
+		return targets
+	}
+
+	// The ranges of the orchestrate blocks as we have to create targets with these ranges
+	// to ensure they are only available within orchestrate blocks
+	ranges := make([]hcl.Range, 0)
+	for _, rule := range record.Meta.OrchestrationRules {
+		ranges = append(ranges, rule.Range)
+	}
+
+	// The names of the existing components in the stack
+	// We use this to offer completions for the component_changes map
+	componentNames := make([]string, 0)
+	for name := range record.Meta.Components {
+		componentNames = append(componentNames, name)
+	}
+
+	for _, rng := range ranges {
```

Comment:
Do we need to store the ranges again when they're already in the shape we need here?

Reply:
Yeah, makes sense! I was unsure whether I would need this multiple time but ended up with just two loops over them 👍